hazelcast/src/main/java/com/hazelcast/multimap/impl/MultiMapContainer.java

```diff
@@ -226,68 +226,74 @@ public int getPartitionId() {
+        if (newValues != null && !newValues.isEmpty()) {
+            MultiMapValue mergedValue = getOrCreateMultiMapValue(mergingEntry.getKey());
+            Collection<MultiMapRecord> records = mergedValue.getCollection(false);
+            createNewMultiMapRecords(records, newValues);
+            if (newValues.equals(mergingEntry.getValue())) {
+                setMergedStatistics(mergingEntry, mergedValue);
             }
+            return mergedValue;
         }
-        return mergedValue;
+        return null;
     }
 
-    private MultiMapValue mergeExistingValue(SplitBrainMergePolicy<Object, MultiMapMergeTypes> mergePolicy,
-                                             SerializationService ss, Data key, MultiMapMergeContainer mergeContainer,
-                                             MultiMapValue existingValue) {
-        boolean isBinary = getConfig().isBinary();
-
+    private MultiMapValue mergeExistingValue(SplitBrainMergePolicy<Collection<Object>, MultiMapMergeTypes> mergePolicy,
+                                             MultiMapMergeTypes mergingEntry, MultiMapValue existingValue,
+                                             SerializationService ss) {
         Collection<MultiMapRecord> existingRecords = existingValue.getCollection(false);
-        int existingHits = existingValue.getHits();
-        for (MultiMapRecord mergeRecord : mergeContainer.getRecords()) {
-            MultiMapMergeTypes mergingEntry = createMergingEntry(ss, mergeContainer, mergeRecord);
-            MultiMapMergeTypes existingEntry = null;
-            MultiMapRecord existingRecord = null;
-            for (MultiMapRecord record : existingRecords) {
-                if (record.getObject().equals(mergeRecord.getObject())) {
-                    existingEntry = createMergingEntry(ss, this, key, record, existingHits);
-                    existingRecord = record;
-                }
-            }
-            Object newValue = mergePolicy.merge(mergingEntry, existingEntry);
-            if (newValue != null && (existingRecord == null || !newValue.equals(existingRecord.getObject()))) {
-                MultiMapRecord newRecord = new MultiMapRecord(nextId(), isBinary ? newValue : ss.toObject(newValue));
-                existingRecords.remove(existingRecord);
-                existingRecords.add(newRecord);
+
+        Data dataKey = mergingEntry.getKey();
+        MultiMapMergeTypes existingEntry = createMergingEntry(ss, this, dataKey, existingRecords, existingValue.getHits());
+        Collection<Object> newValues = mergePolicy.merge(mergingEntry, existingEntry);
+        if (newValues == null || newValues.isEmpty()) {
+            existingRecords.clear();
+            multiMapValues.remove(dataKey);
+        } else if (!newValues.equals(existingRecords)) {
+            existingRecords.clear();
+            createNewMultiMapRecords(existingRecords, newValues);
+            if (newValues.equals(mergingEntry.getValue())) {
```

Comment:
Same here for existing values (https://github.com/hazelcast/hazelcast/issues/12805)